crates/ruff_python_formatter/src/statement/stmt_with.rs

```diff
@@ -98,14 +108,52 @@ impl FormatNodeRule<StmtWith> for FormatStmtWith {
     }
 }
 
+/// Returns `true` if the `with` items should be parenthesized, if at least one item expands.
+///
+/// Black parenthesizes `with` items if there's more than one item and they're already
+/// parenthesized, _or_ there's a single item with a trailing comma.
+fn should_parenthesize(with: &StmtWith, context: &PyFormatContext) -> FormatResult<bool> {
+    if has_magic_trailing_comma(with, context) {
+        return Ok(true);
+    }
+
+    if with.items.len() > 1 && are_with_items_parenthesized(with, context)? {
+        return Ok(true);
+    }
+
+    Ok(false)
+}
+
+fn has_magic_trailing_comma(with: &StmtWith, context: &PyFormatContext) -> bool {
+    let Some(last_item) = with.items.last() else {
+        return false;
+    };
+
+    let first_token = SimpleTokenizer::new(
+        context.source(),
+        TextRange::new(last_item.end(), with.end()),
+    )
+    .skip_trivia()
+    // Skip over any closing parentheses belonging to the expression
+    .find(|token| token.kind() != SimpleTokenKind::RParen);
+
+    matches!(
+        first_token,
+        Some(SimpleToken {
+            kind: SimpleTokenKind::Comma,
+            ..
+        })
+    )
+}
```

Comment:
This must respect the `f.options().magic_trailing_comma` option. It could be worth adding an extra test file that sets the option to ignore similar to `skip_magic_trailing_comma.py`

Could we reuse/extract the logic from 

https://github.com/astral-sh/ruff/blob/b1c4c7be69d796186e0a75e0e88955b489343d9a/crates/ruff_python_formatter/src/builders.rs#L197-L216

Reply:
Good call.